lib/kvs/kvs_cursor.c

```diff
@@ -1100,32 +1100,58 @@ ikvs_cursor_seek(struct kvs_cursor_impl *cursor, const void *key, size_t klen)
     return err;
 }
 
-static merr_t
-ikvs_cursor_kv_copy(struct kvs_cursor_impl *cursor, struct kvs_kvtuple *kvt)
+const void *
+kvs_cursor_key_copy(struct hse_kvs_cursor *cursor, const void *kbuf, size_t kbufsz, size_t *klen)
 {
-    struct kvs_vtuple *vt = &cursor->kci_elem_last.kce_vt;
-    uint               clen = cursor->kci_elem_last.kce_complen;
-    void *             vbuf = cursor->kci_buf + HSE_KVS_KEY_LEN_MAX;
+    struct kvs_cursor_impl *cur = cursor_h2r(cursor);
+    void                   *buf = (void *)kbuf;
+    size_t                  bufsz = kbufsz;
+    uint                    keylen;
+    const void             *key;
+
+    if (!buf) {
+        buf = (void *)cur->kci_buf;
+        bufsz = HSE_KVS_KEY_LEN_MAX;
+    }
+
+    key = key_obj_copy(buf, bufsz, &keylen, cur->kci_last);
+    *klen = keylen;
+
+    return key;
+}
+
+const void *
+kvs_cursor_val_copy(struct hse_kvs_cursor *cursor, const void *vbuf, size_t vbufsz, size_t *vlen)
+{
+    struct kvs_cursor_impl *cur = cursor_h2r(cursor);
+
+    struct kvs_vtuple *vt = &cur->kci_elem_last.kce_vt;
+    uint               clen = cur->kci_elem_last.kce_complen;
+    void              *buf = (void *)vbuf;
+    size_t             bufsz = vbufsz;
     merr_t             err = 0;
 
-    kvt->kvt_key.kt_data = key_obj_copy(
-        cursor->kci_buf, HSE_KVS_KEY_LEN_MAX, (uint *)&kvt->kvt_key.kt_len, cursor->kci_last);
+    if (!buf) {
+        buf = cur->kci_buf + HSE_KVS_KEY_LEN_MAX;
+        bufsz = HSE_KVS_VALUE_LEN_MAX;
+    }
+
+    *vlen = vt->vt_xlen;
 
     if (clen) {
         uint outlen;
 
-        err = compress_lz4_ops.cop_decompress(vt->vt_data, clen, vbuf, HSE_KVS_VALUE_LEN_MAX, &outlen);
+        err = compress_lz4_ops.cop_decompress(vt->vt_data, clen, buf, bufsz, &outlen);
         if (ev(err))
-            return err;
+            return NULL;
 
-        if (ev(outlen != vt->vt_xlen))
-            return merr(EBUG);
+        if (ev(outlen != min_t(u64, vt->vt_xlen, bufsz)))
+            return NULL;
     } else {
-        memcpy(vbuf, vt->vt_data, vt->vt_xlen);
+        memcpy(buf, vt->vt_data, min_t(u64, vt->vt_xlen, bufsz));
     }
 
-    kvs_vtuple_init(&kvt->kvt_value, vbuf, vt->vt_xlen);
-    return err;
+    return buf;
```

Comment:
Not a bug since cursors do not pack data into the upper 32-bits of vt_xlen.  I don't like the fact that vt_xlen is used in different ways depending on context, but it's not a bug that needs to be fixed in this PR.

Reply:
Ok, so future we we make it a union with a different name for use by cursors...  Or some such, too confusing as-is...